src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -91,33 +91,50 @@ class UserService(val directoryDAO: DirectoryDAO, val googleDirectoryDAO: Google
-        _ <- Future.traverse(petServiceAccountConfig.serviceAccountActors) { email =>
-          googleIamDAO.addServiceAccountActorRoleForUser(petServiceAccountConfig.googleProject, petServiceAccount.email, email)
-        }
-        _ <- directoryDAO.addPetServiceAccountToUser(user.id, petServiceAccount.email)
-      } yield petServiceAccount.email
-    }
-
     directoryDAO.getPetServiceAccountForUser(user.id).flatMap {
       case Some(email) => Future.successful(email)
-      case None => create().andThen { case Failure(_) =>
-        // If anything fails with pet service account creation, clean up
-        // any created resources asynchronously, to ensure we don't end
-        // up with orphaned pets.
-        Future.sequence(Seq(
-          directoryDAO.removePetServiceAccountFromUser(user.id),
-          googleIamDAO.removeServiceAccount(petServiceAccountConfig.googleProject, petSa))
-        ).failed.foreach { e =>
-          logger.warn(s"Error occurred cleaning up pet service account [$petSa] [$petSaDisplayName]", e)
+      case None =>
+        // First create the service account in Google, which generates a unique id and email
+        googleIamDAO.createServiceAccount(petServiceAccountConfig.googleProject, petSa, petSaDisplayName).flatMap { petServiceAccount =>
+          // Set up the service account with the necessary permissions
+          setUpServiceAccount(user, petServiceAccount) andThen { case Failure(_) =>
+            // If anything fails with setup, clean up any created resources to ensure we don't end up with orphaned pets.
+            removePetServiceAccount(user, petServiceAccount).failed.foreach { e =>
+              logger.warn(s"Error occurred cleaning up pet service account [$petSa] [$petSaDisplayName]", e)
+            }
         }
       }
     }
   }
 
+  private def setUpServiceAccount(user: WorkbenchUser, petServiceAccount: WorkbenchUserServiceAccount): Future[WorkbenchUserServiceAccountEmail] = {
+    for {
+      // add the pet service account to the user's proxy group
+      _ <- googleDirectoryDAO.addMemberToGroup(WorkbenchGroupEmail(toProxyFromUser(user.id.value)), petServiceAccount.email)
+      // add Service Account Actor role to the configured emails so they can assume the identity of the pet service account
+      _ <- Future.traverse(petServiceAccountConfig.serviceAccountUsers) { email =>
+        googleIamDAO.addServiceAccountUserRoleForUser(petServiceAccountConfig.googleProject, petServiceAccount.email, email)
+      }
+      // add the pet service account attribute to the user's LDAP record
+      _ <- directoryDAO.addPetServiceAccountToUser(user.id, petServiceAccount.email)
+      // create an additional LDAP record for the pet service account itself (in a different organizational unit than the user)
+      _ <- directoryDAO.createPetServiceAccount(petServiceAccount)
+      // add the pet service account to the 'enabled users' group
+      _ <- directoryDAO.enableUser(user.id)
```

Comment:
is this an avenue through which a disabled user can inadvertently be enabled? But maybe they can't even call this if they are disabled.

Reply:
I am not certain if this can be called for a disabled user but seems risky. I think we need to only enable the pet SA if the user is enabled.

Reply:
I _think_ the user will be blocked at the proxy level if disabled.. but I can change this to enable only the pet as you suggested.

Reply:
Think I found another bug -- `enableUser`/`disableUser` is add/removing the user email from the proxy group -- but NOT removing the pet email. So as it stands the pet email will remain in the proxy group after disabling the user. Fixing that as well.